File: go/capabilities_test.go
// Tests for the capability announcement + filter surface (Stage G-2).
//
// Each node self-indexes its own announcement, so single-node
// round-trip covers the dict→core conversion plus filter predicate
// end-to-end. Multi-node propagation is covered by the Rust
// integration suite (`tests/capability_broadcast.rs`).

package net

import (
	"slices"
	"testing"
)

func newMeshForCaps(t *testing.T) *MeshNode {
	t.Helper()
	addr := reserveLocalUDPPort(t)
	m, err := NewMeshNode(MeshConfig{BindAddr: addr, PskHex: meshPsk})
	if err != nil {
		t.Fatalf("new mesh: %v", err)
	}
	return m
}

// ---------------------------------------------------------------------------
// Self-match round-trip
// ---------------------------------------------------------------------------

func TestAnnounce_ThenFind_SelfMatchesOnTag(t *testing.T) {
	m := newMeshForCaps(t)
	defer m.Shutdown()

	if err := m.AnnounceCapabilities(CapabilitySet{Tags: []string{"gpu", "prod"}}); err != nil {
		t.Fatalf("announce: %v", err)
	}
	peers, err := m.FindNodes(CapabilityFilter{RequireTags: []string{"gpu"}})
	if err != nil {
		t.Fatalf("find_nodes: %v", err)
	}
	if !slices.Contains(peers, m.NodeID()) {
		t.Fatalf("own node id missing from find_nodes result: %v", peers)
	}
}

func TestFindNodes_EmptyWhenFilterMismatches(t *testing.T) {
	m := newMeshForCaps(t)
	defer m.Shutdown()

	if err := m.AnnounceCapabilities(CapabilitySet{Tags: []string{"cpu"}}); err != nil {
		t.Fatalf("announce: %v", err)
	}
	peers, err := m.FindNodes(CapabilityFilter{RequireTags: []string{"gpu"}})
	if err != nil {
		t.Fatalf("find_nodes: %v", err)
	}
	if len(peers) != 0 {
		t.Fatalf("expected empty peer list, got %v", peers)
	}
}

func TestFindNodes_WithoutAnnouncement_IsEmpty(t *testing.T) {
	m := newMeshForCaps(t)
	defer m.Shutdown()

	peers, err := m.FindNodes(CapabilityFilter{RequireTags: []string{"anything"}})
	if err != nil {
		t.Fatalf("find_nodes: %v", err)
	}
	if len(peers) != 0 {
		t.Fatalf("expected empty peer list, got %v", peers)
	}
}

// ---------------------------------------------------------------------------
// Hardware / model / tool filters
// ---------------------------------------------------------------------------

func TestHardwareAndGpuFilter_Matches(t *testing.T) {
	m := newMeshForCaps(t)
	defer m.Shutdown()

	err := m.AnnounceCapabilities(CapabilitySet{
		Hardware: &HardwareCaps{
			CPUCores: 16,
			MemoryGB: 64,
			GPU:      &GPUInfo{Vendor: "nvidia", Model: "h100", VRAMGB: 80},
		},
		Tags: []string{"gpu"},
	})
	if err != nil {
		t.Fatalf("announce: %v", err)
	}

	peers, err := m.FindNodes(CapabilityFilter{
		RequireGPU:  true,
		GPUVendor:   "nvidia",
		MinVRAMGB:   40,
		MinMemoryGB: 32,
	})
	if err != nil {
		t.Fatalf("find_nodes: %v", err)
	}
	if !slices.Contains(peers, m.NodeID()) {
		t.Fatalf("own id missing: %v", peers)
	}

	strict, err := m.FindNodes(CapabilityFilter{MinVRAMGB: 200})
	if err != nil {
		t.Fatalf("find_nodes strict: %v", err)
	}
	if len(strict) != 0 {
		t.Fatalf("expected strict filter to miss, got %v", strict)
	}
}

func TestModelAndToolFilter_Matches(t *testing.T) {
	m := newMeshForCaps(t)
	defer m.Shutdown()

	err := m.AnnounceCapabilities(CapabilitySet{
		Models: []ModelCaps{{
			ModelID:       "llama-3.1-70b",
			Family:        "llama",
			ParametersBx10: 700,
			ContextLength: 128_000,
			Modalities:    []string{"text", "code"},
		}},
		Tools: []ToolCaps{{ToolID: "sql_exec", Name: "SQL Exec"}},
	})
	if err != nil {
		t.Fatalf("announce: %v", err)
	}

	peers, err := m.FindNodes(CapabilityFilter{RequireModels: []string{"llama-3.1-70b"}})
	if err != nil || !slices.Contains(peers, m.NodeID()) {
		t.Fatalf("model filter missed own node: err=%v peers=%v", err, peers)
	}
	peers, err = m.FindNodes(CapabilityFilter{RequireTools: []string{"sql_exec"}})
	if err != nil || !slices.Contains(peers, m.NodeID()) {
		t.Fatalf("tool filter missed own node: err=%v peers=%v", err, peers)
	}
	peers, err = m.FindNodes(CapabilityFilter{
		RequireModalities: []string{"code"},
		MinContextLength:  100_000,
	})
	if err != nil || !slices.Contains(peers, m.NodeID()) {
		t.Fatalf("modality filter missed own node: err=%v peers=%v", err, peers)
	}

	missing, err := m.FindNodes(CapabilityFilter{RequireModels: []string{"missing"}})
	if err != nil {
		t.Fatalf("find_nodes missing-model query: %v", err)
	}
	if len(missing) != 0 {
		t.Fatalf("expected no match, got %v", missing)
	}
}

func TestEmptyAnnouncement_StillSelfIndexes(t *testing.T) {
	m := newMeshForCaps(t)
	defer m.Shutdown()

	if err := m.AnnounceCapabilities(CapabilitySet{}); err != nil {
		t.Fatalf("announce: %v", err)
	}
	peers, err := m.FindNodes(CapabilityFilter{})
	if err != nil {
		t.Fatalf("find_nodes: %v", err)
	}
	if !slices.Contains(peers, m.NodeID()) {
		t.Fatalf("expected own node indexed, got %v", peers)
	}
}

// ---------------------------------------------------------------------------
// Vendor normalization helper
// ---------------------------------------------------------------------------

func TestNormalizeGPUVendor(t *testing.T) {
	cases := []struct {
		raw, want string
	}{
		{"NVIDIA", "nvidia"},
		{"Nvidia", "nvidia"},
		{"amd", "amd"},
		{"Apple", "apple"},
		{"qualcomm", "qualcomm"},
		{"intel", "intel"},
		{"bogus", "unknown"},
		{"", "unknown"},
	}
	for _, c := range cases {
		got, err := NormalizeGPUVendor(c.raw)
		if err != nil {
			t.Fatalf("normalize %q: %v", c.raw, err)
		}
		if got != c.want {
			t.Fatalf("normalize %q: want %q, got %q", c.raw, c.want, got)
		}
	}
}

// ---------------------------------------------------------------------------
// Scoped discovery (reserved scope:* tags)
// ---------------------------------------------------------------------------

func TestFindNodesScoped_TenantTagFiltersOutOtherTenants(t *testing.T) {
	m := newMeshForCaps(t)
	defer m.Shutdown()

	// Announce ourselves under tenant `oem-123`.
	caps := CapabilitySet{
		Tags: []string{"model:llama3-70b", "scope:tenant:oem-123"},
	}
	if err := m.AnnounceCapabilities(caps); err != nil {
		t.Fatalf("announce: %v", err)
	}
	filter := CapabilityFilter{RequireTags: []string{"model:llama3-70b"}}

	// Tenant("oem-123"): includes us.
	peers, err := m.FindNodesScoped(filter, ScopeFilter{Kind: "tenant", Tenant: "oem-123"})
	if err != nil {
		t.Fatalf("find_nodes_scoped tenant=oem-123: %v", err)
	}
	if !slices.Contains(peers, m.NodeID()) {
		t.Fatalf("own node missing under tenant=oem-123: %v", peers)
	}

	// Tenant("corp-acme"): excludes us — different tenant.
	peers, err = m.FindNodesScoped(filter, ScopeFilter{Kind: "tenant", Tenant: "corp-acme"})
	if err != nil {
		t.Fatalf("find_nodes_scoped tenant=corp-acme: %v", err)
	}
	if slices.Contains(peers, m.NodeID()) {
		t.Fatalf("own node leaked into tenant=corp-acme: %v", peers)
	}

	// Any: includes us (no SubnetLocal exclusion applies).
	peers, err = m.FindNodesScoped(filter, ScopeFilter{Kind: "any"})
	if err != nil {
		t.Fatalf("find_nodes_scoped any: %v", err)
	}
	if !slices.Contains(peers, m.NodeID()) {
		t.Fatalf("own node missing under any: %v", peers)
	}

	// GlobalOnly: excludes us — we have a tenant tag.
	peers, err = m.FindNodesScoped(filter, ScopeFilter{Kind: "global_only"})
	if err != nil {
		t.Fatalf("find_nodes_scoped global_only: %v", err)
	}
	if slices.Contains(peers, m.NodeID()) {
		t.Fatalf("tenant-tagged node leaked into global_only: %v", peers)
	}
}

func TestFindNodesScoped_GlobalNodeVisibleToTenantQuery(t *testing.T) {
	// A node without any `scope:*` tag is `Global` — visible to
	// tenant queries by design (permissive default; matches the v1
	// behaviour for nodes that don't opt in to scope tagging).
	m := newMeshForCaps(t)
	defer m.Shutdown()

	if err := m.AnnounceCapabilities(CapabilitySet{Tags: []string{"gpu"}}); err != nil {
		t.Fatalf("announce: %v", err)
	}
	peers, err := m.FindNodesScoped(
		CapabilityFilter{RequireTags: []string{"gpu"}},
		ScopeFilter{Kind: "tenant", Tenant: "oem-123"},
	)
	if err != nil {
		t.Fatalf("find_nodes_scoped: %v", err)
	}
	if !slices.Contains(peers, m.NodeID()) {
		t.Fatalf("untagged (Global) node should match tenant query, got %v", peers)
	}
}

// ---------------------------------------------------------------------------
// FindBestNode — scored placement
// ---------------------------------------------------------------------------

func TestFindBestNode_SelfMatchesOnFilter(t *testing.T) {
	m := newMeshForCaps(t)
	defer m.Shutdown()

	if err := m.AnnounceCapabilities(CapabilitySet{
		Hardware: &HardwareCaps{
			MemoryGB: 64,
			GPU:      &GPUInfo{Vendor: "nvidia", Model: "h100", VRAMGB: 80},
		},
		Tags: []string{"gpu"},
	}); err != nil {
		t.Fatalf("announce: %v", err)
	}
	// Filter that matches us; weights aren't load-bearing for a
	// single-candidate set but exercise the scoring path.
	req := CapabilityRequirement{
		Filter:           CapabilityFilter{RequireGPU: true, MinVRAMGB: 40},
		PreferMoreVRAM:   1.0,
		PreferMoreMemory: 0.5,
	}
	nodeID, ok, err := m.FindBestNode(req)
	if err != nil {
		t.Fatalf("find_best_node: %v", err)
	}
	if !ok {
		t.Fatalf("expected a match, got none")
	}
	if nodeID != m.NodeID() {
		t.Fatalf("expected own node id %d, got %d", m.NodeID(), nodeID)
	}
}

func TestFindBestNode_NoMatchReturnsFalseNotError(t *testing.T) {
	m := newMeshForCaps(t)
	defer m.Shutdown()

	if err := m.AnnounceCapabilities(CapabilitySet{Tags: []string{"cpu"}}); err != nil {
		t.Fatalf("announce: %v", err)
	}
	req := CapabilityRequirement{
		Filter: CapabilityFilter{RequireGPU: true},
	}
	nodeID, ok, err := m.FindBestNode(req)
	if err != nil {
		t.Fatalf("find_best_node: %v", err)
	}
	if ok {
		t.Fatalf("expected no match, got node %d", nodeID)
	}
	if nodeID != 0 {
		t.Fatalf("expected nodeID=0 on miss, got %d", nodeID)
	}
}

// Regression: P2 (Cubic) — empty-string sanitization on
// `Tenants` / `Regions` lists. Unsanitized input like `[""]` used
// to flow through to a `Tenants([""])` filter, which matches no
// real tenant and silently narrows results to Global candidates.
// Fix: drop empties; fall back to Any when cleaned list is empty.

func TestFindNodesScoped_TenantsEmptyListFallsBackToAny(t *testing.T) {
	m := newMeshForCaps(t)
	defer m.Shutdown()

	// Tenant-tagged provider — without sanitization, a
	// `tenants: [""]` query would NOT return this node, and
	// would NOT return any Global node either.
	if err := m.AnnounceCapabilities(CapabilitySet{
		Tags: []string{"gpu", "scope:tenant:oem-123"},
	}); err != nil {
		t.Fatalf("announce: %v", err)
	}
	filter := CapabilityFilter{RequireTags: []string{"gpu"}}

	// `[""]` sanitizes to Any → matches own node.
	peers, err := m.FindNodesScoped(filter, ScopeFilter{Kind: "tenants", Tenants: []string{""}})
	if err != nil {
		t.Fatalf("find_nodes_scoped tenants=[\"\"]: %v", err)
	}
	if !slices.Contains(peers, m.NodeID()) {
		t.Fatalf("tenants=[\"\"] must fall back to Any (P2 regression); got %v", peers)
	}

	// `nil` / empty list also falls back to Any.
	peers, err = m.FindNodesScoped(filter, ScopeFilter{Kind: "tenants", Tenants: []string{}})
	if err != nil {
		t.Fatalf("find_nodes_scoped tenants=[]: %v", err)
	}
	if !slices.Contains(peers, m.NodeID()) {
		t.Fatalf("tenants=[] must fall back to Any; got %v", peers)
	}
}

func TestFindNodesScoped_TenantsPartialCleanDropsEmpties(t *testing.T) {
	m := newMeshForCaps(t)
	defer m.Shutdown()

	if err := m.AnnounceCapabilities(CapabilitySet{
		Tags: []string{"gpu", "scope:tenant:oem-123"},
	}); err != nil {
		t.Fatalf("announce: %v", err)
	}
	filter := CapabilityFilter{RequireTags: []string{"gpu"}}

	// `["", "oem-123"]` sanitizes to `Tenants(["oem-123"])`.
	peers, err := m.FindNodesScoped(filter, ScopeFilter{
		Kind:    "tenants",
		Tenants: []string{"", "oem-123"},
	})
	if err != nil {
		t.Fatalf("find_nodes_scoped: %v", err)
	}
	if !slices.Contains(peers, m.NodeID()) {
		t.Fatalf("partial-clean must keep \"oem-123\" filter; got %v", peers)
	}

	// `["", "corp-acme"]` excludes us (different tenant).
	peers, err = m.FindNodesScoped(filter, ScopeFilter{
		Kind:    "tenants",
		Tenants: []string{"", "corp-acme"},
	})
	if err != nil {
		t.Fatalf("find_nodes_scoped: %v", err)
	}
	if slices.Contains(peers, m.NodeID()) {
		t.Fatalf("partial-clean with non-matching tenant must exclude us; got %v", peers)
	}
}

func TestFindNodesScoped_RegionsEmptyListFallsBackToAny(t *testing.T) {
	m := newMeshForCaps(t)
	defer m.Shutdown()

	if err := m.AnnounceCapabilities(CapabilitySet{
		Tags: []string{"relay-capable", "scope:region:eu-west"},
	}); err != nil {
		t.Fatalf("announce: %v", err)
	}
	filter := CapabilityFilter{RequireTags: []string{"relay-capable"}}

	peers, err := m.FindNodesScoped(filter, ScopeFilter{Kind: "regions", Regions: []string{""}})
	if err != nil {
		t.Fatalf("find_nodes_scoped regions=[\"\"]: %v", err)
	}
	if !slices.Contains(peers, m.NodeID()) {
		t.Fatalf("regions=[\"\"] must fall back to Any (P2 regression); got %v", peers)
	}

	peers, err = m.FindNodesScoped(filter, ScopeFilter{Kind: "regions", Regions: []string{}})
	if err != nil {
		t.Fatalf("find_nodes_scoped regions=[]: %v", err)
	}
	if !slices.Contains(peers, m.NodeID()) {
		t.Fatalf("regions=[] must fall back to Any; got %v", peers)
	}
}

func TestFindBestNodeScoped_SelfMatchesUnderTenantScope(t *testing.T) {
	m := newMeshForCaps(t)
	defer m.Shutdown()

	if err := m.AnnounceCapabilities(CapabilitySet{
		Tags: []string{"model:llama3-70b", "scope:tenant:oem-123"},
	}); err != nil {
		t.Fatalf("announce: %v", err)
	}

	req := CapabilityRequirement{
		Filter: CapabilityFilter{RequireTags: []string{"model:llama3-70b"}},
	}

	// Tenant("oem-123") — own node matches.
	nodeID, ok, err := m.FindBestNodeScoped(req, ScopeFilter{Kind: "tenant", Tenant: "oem-123"})
	if err != nil {
		t.Fatalf("find_best_node_scoped tenant=oem-123: %v", err)
	}
	if !ok || nodeID != m.NodeID() {
		t.Fatalf("expected own node id %d under matching tenant, got (%d, %v)",
			m.NodeID(), nodeID, ok)
	}

	// Tenant("corp-acme") — no match (different tenant; tenant-tagged
	// nodes are excluded from non-matching tenant queries).
	nodeID, ok, err = m.FindBestNodeScoped(req, ScopeFilter{Kind: "tenant", Tenant: "corp-acme"})
	if err != nil {
		t.Fatalf("find_best_node_scoped tenant=corp-acme: %v", err)
	}
	if ok {
		t.Fatalf("expected no match under non-matching tenant, got node %d", nodeID)
	}
}

// ---------------------------------------------------------------------------
// Constructor kwargs (capability_gc + signed)
// ---------------------------------------------------------------------------

func TestMeshConfig_AcceptsCapabilityKwargs(t *testing.T) {
	addr := reserveLocalUDPPort(t)
	m, err := NewMeshNode(MeshConfig{
		BindAddr:                  addr,
		PskHex:                    meshPsk,
		CapabilityGCIntervalMs:    120_000,
		RequireSignedCapabilities: true,
	})
	if err != nil {
		t.Fatalf("new mesh with cap kwargs: %v", err)
	}
	defer m.Shutdown()
	if m.NodeID() == 0 {
		t.Fatal("node id unset")
	}
}

File: net/crates/net/sdk-ts/test/capabilities.test.ts
// Integration tests for the capability-broadcast surface on
// `MeshNode`. Exercises announce → find round-trip, self-index,
// late-joiner push, and the flattened POJO conversions — mirrors the
// Rust `tests/capability_broadcast.rs` suite.

import { afterEach, describe, expect, it } from 'vitest';

import type { CapabilityFilter, CapabilitySet } from '../src/capabilities';
import { MeshNode } from '../src/mesh';

const PSK = '42'.repeat(32);

let portSeed = 29_400;
function nextPortPair(): [string, string] {
  const a = portSeed++;
  const b = portSeed++;
  return [`127.0.0.1:${a}`, `127.0.0.1:${b}`];
}

async function handshake(a: MeshNode, b: MeshNode, bAddr: string): Promise<void> {
  const bPub = b.publicKey();
  const aId = a.nodeId();
  const bId = b.nodeId();
  await Promise.all([
    b.accept(aId),
    (async () => {
      await new Promise((r) => setTimeout(r, 50));
      await a.connect(bAddr, bPub, bId);
    })(),
  ]);
  await a.start();
  await b.start();
}

async function pair(): Promise<{ a: MeshNode; b: MeshNode; bAddr: string }> {
  const [aAddr, bAddr] = nextPortPair();
  const a = await MeshNode.create({ bindAddr: aAddr, psk: PSK });
  const b = await MeshNode.create({ bindAddr: bAddr, psk: PSK });
  await handshake(a, b, bAddr);
  return { a, b, bAddr };
}

async function waitUntil(fn: () => boolean, timeoutMs = 2_000): Promise<boolean> {
  const deadline = Date.now() + timeoutMs;
  while (Date.now() < deadline) {
    if (fn()) return true;
    await new Promise((r) => setTimeout(r, 25));
  }
  return fn();
}

const nodes: MeshNode[] = [];
afterEach(async () => {
  while (nodes.length > 0) {
    const n = nodes.pop()!;
    try {
      await n.shutdown();
    } catch {
      // Ignore — test may have already closed the node.
    }
  }
});

describe('MeshNode capabilities', () => {
  it('self-matches on its own announcement (single node)', async () => {
    const a = await MeshNode.create({ bindAddr: '127.0.0.1:0', psk: PSK });
    nodes.push(a);

    await a.announceCapabilities({ tags: ['gpu', 'inference'] });

    const hits = a.findNodes({ requireTags: ['gpu'] });
    expect(hits).toContain(a.nodeId());
  });

  it('returns empty for a non-matching filter', async () => {
    const a = await MeshNode.create({ bindAddr: '127.0.0.1:0', psk: PSK });
    nodes.push(a);
    await a.announceCapabilities({ tags: ['gpu'] });

    expect(a.findNodes({ requireTags: ['nope'] })).toEqual([]);
  });

  it('announce → find propagates across the handshake (two nodes)', async () => {
    const { a, b } = await pair();
    nodes.push(a, b);

    await a.announceCapabilities({
      hardware: {
        cpuCores: 16,
        memoryGb: 64,
        gpu: { vendor: 'nvidia', model: 'RTX 4090', vramGb: 24 },
      },
      tags: ['gpu', 'inference'],
    });

    const aId = a.nodeId();
    const filter: CapabilityFilter = { requireGpu: true, minVramMb: 16_384 };

    const arrived = await waitUntil(() => b.findNodes(filter).includes(aId));
    expect(arrived).toBe(true);
  });

  it('late joiner receives session-open push', async () => {
    const [aAddr, bAddr] = nextPortPair();
    const a = await MeshNode.create({ bindAddr: aAddr, psk: PSK });
    nodes.push(a);

    // A announces *before* B exists.
    await a.announceCapabilities({ tags: ['preannounced'] });

    const b = await MeshNode.create({ bindAddr: bAddr, psk: PSK });
    nodes.push(b);
    await handshake(a, b, bAddr);

    const aId = a.nodeId();
    const arrived = await waitUntil(() =>
      b.findNodes({ requireTags: ['preannounced'] }).includes(aId),
    );
    expect(arrived).toBe(true);
  });

  it('round-trips a complex POJO (hardware + software + models + tools)', async () => {
    const a = await MeshNode.create({ bindAddr: '127.0.0.1:0', psk: PSK });
    nodes.push(a);

    const caps: CapabilitySet = {
      hardware: {
        cpuCores: 32,
        cpuThreads: 64,
        memoryGb: 128,
        gpu: {
          vendor: 'nvidia',
          model: 'H100',
          vramGb: 80,
          computeUnits: 132,
        },
        storageGb: 4000n,
      },
      software: {
        os: 'linux',
        osVersion: '6.5.0',
        runtimes: [
          ['python', '3.12.4'],
          ['node', '22.3.0'],
        ],
        cudaVersion: '12.4',
      },
      models: [
        {
          modelId: 'llama-3.1-70b',
          family: 'llama',
          parametersBX10: 700,
          contextLength: 131_072,
          modalities: ['text', 'code'],
          loaded: true,
        },
      ],
      tools: [{ toolId: 'web_search', name: 'web search', stateless: true }],
      tags: ['prod', 'us-east'],
      limits: { maxConcurrentRequests: 32 },
    };
    await a.announceCapabilities(caps);

    // Each require* dimension resolves back to self.
    expect(a.findNodes({ requireTags: ['prod'] })).toContain(a.nodeId());
    expect(a.findNodes({ requireModels: ['llama-3.1-70b'] })).toContain(a.nodeId());
    expect(a.findNodes({ requireTools: ['web_search'] })).toContain(a.nodeId());
    expect(a.findNodes({ requireModalities: ['code'] })).toContain(a.nodeId());
    expect(a.findNodes({ gpuVendor: 'nvidia', minVramMb: 40_000 })).toContain(a.nodeId());
  });

  it('drops expired entries after TTL + a GC sweep', async () => {
    // Short interval so the test completes fast. The announcement
    // TTL is 1 s; a 200 ms GC tick means two-to-three sweeps happen
    // before the 1.5 s re-query.
    //
    // `announceCapabilities` takes no TTL override from the TS
    // surface yet — the core default is 5 min — so we'd normally
    // wait minutes. Use `capabilityGcIntervalMs` to speed GC and
    // reach the core `announce_capabilities_with(..., ttl)` via the
    // plain `announceCapabilities` path on the Rust side (which
    // uses 5 min). That TTL is too long here, so we skip the
    // "eventually empty" assertion and instead verify:
    //   (a) the announcement IS indexed (positive path), and
    //   (b) the `capabilityGcIntervalMs` knob is accepted.
    //
    // Full TTL expiry is covered by
    // `tests/capability_broadcast.rs::announcement_expires_after_ttl`
    // where `announce_capabilities_with(caps, 1s, false)` is
    // available at the core Rust layer.
    const a = await MeshNode.create({
      bindAddr: '127.0.0.1:0',
      psk: PSK,
      capabilityGcIntervalMs: 200,
    });
    nodes.push(a);
    await a.announceCapabilities({ tags: ['gc-smoke'] });
    expect(a.findNodes({ requireTags: ['gc-smoke'] })).toContain(a.nodeId());
  });

  it('accepts the requireSignedCapabilities knob without breaking the local path', async () => {
    // `announceCapabilities` path stamps no signature on the wire
    // (signing binds to Stage E), so with
    // `requireSignedCapabilities = true` on a receiver, a direct
    // unsigned announcement would be dropped. We can't easily
    // two-node-test that over the TS boundary (no TS-side wire
    // control), so this is a config-plumbing smoke: the option is
    // accepted and the local self-index path still works because
    // `announce_capabilities` indexes locally before sending.
    const a = await MeshNode.create({
      bindAddr: '127.0.0.1:0',
      psk: PSK,
      requireSignedCapabilities: true,
    });
    nodes.push(a);
    await a.announceCapabilities({ tags: ['local-only'] });
    expect(a.findNodes({ requireTags: ['local-only'] })).toContain(a.nodeId());
  });

  // Scope-tag discovery — the NAPI layer has unique plumbing
  // (`ScopeFilterJs` → `ScopeFilterOwned` → `with_scope_filter`
  // borrow trampoline). Exercise it end-to-end with a single-node
  // self-match; the underlying matching logic is covered by the
  // Rust unit + integration suites.
  it('findNodesScoped: tenant-tagged self matches under matching tenant', async () => {
    const a = await MeshNode.create({ bindAddr: '127.0.0.1:0', psk: PSK });
    nodes.push(a);

    await a.announceCapabilities({
      tags: ['model:llama3-70b', 'scope:tenant:oem-123'],
    });

    const filter: CapabilityFilter = { requireTags: ['model:llama3-70b'] };

    // Matching tenant — own node id appears.
    expect(a.findNodesScoped(filter, { kind: 'tenant', tenant: 'oem-123' })).toContain(
      a.nodeId(),
    );

    // Non-matching tenant — own node id is excluded (tenant-tagged
    // peer is invisible to other-tenant queries).
    expect(a.findNodesScoped(filter, { kind: 'tenant', tenant: 'corp-acme' })).not.toContain(
      a.nodeId(),
    );

    // GlobalOnly — tenant-tagged peer also excluded.
    expect(a.findNodesScoped(filter, { kind: 'globalOnly' })).not.toContain(a.nodeId());
  });

  it('findNodesScoped: untagged Global peer remains visible to tenant queries', async () => {
    // The permissive default — a node that doesn't tag itself stays
    // discoverable under tenant-scoped queries. Locks in the v1
    // backwards-compat behaviour through the JS-side scope filter.
    const a = await MeshNode.create({ bindAddr: '127.0.0.1:0', psk: PSK });
    nodes.push(a);

    await a.announceCapabilities({ tags: ['gpu'] });

    expect(
      a.findNodesScoped({ requireTags: ['gpu'] }, { kind: 'tenant', tenant: 'oem-123' }),
    ).toContain(a.nodeId());
  });

  it('findNodesScoped: regions list variant marshals correctly through NAPI', async () => {
    // Multi-element variants (`tenants` / `regions`) take a separate
    // path in `with_scope_filter` because they need an intermediate
    // `Vec<&str>` whose lifetime outlives the borrow. This test
    // hits that borrow trampoline end-to-end.
    const a = await MeshNode.create({ bindAddr: '127.0.0.1:0', psk: PSK });
    nodes.push(a);

    await a.announceCapabilities({
      tags: ['relay-capable', 'scope:region:eu-west'],
    });

    const filter: CapabilityFilter = { requireTags: ['relay-capable'] };

    // Multi-region list including ours — match.
    expect(
      a.findNodesScoped(filter, { kind: 'regions', regions: ['us-east', 'eu-west'] }),
    ).toContain(a.nodeId());

    // Multi-region list excluding ours — no match.
    expect(
      a.findNodesScoped(filter, { kind: 'regions', regions: ['us-east', 'ap-south'] }),
    ).not.toContain(a.nodeId());
  });

  // Regression: P2 (Cubic) — empty-string sanitization on
  // `tenants` / `regions` lists. Unsanitized input like `[""]`
  // used to flow through to a `Tenants([""])` filter, which
  // matches no real tenant and silently narrows results to
  // Global candidates. Fix: drop empties; if list is empty
  // after cleaning, fall back to Any.
  it('findNodesScoped: tenants list with only empty strings falls back to Any', async () => {
    const a = await MeshNode.create({ bindAddr: '127.0.0.1:0', psk: PSK });
    nodes.push(a);

    // Tenant-tagged provider — without sanitization, a
    // `tenants: [""]` query would *not* return this node
    // (empty string never matches "oem-123") and would NOT
    // return Global nodes either (none exist here).
    await a.announceCapabilities({
      tags: ['gpu', 'scope:tenant:oem-123'],
    });

    // After sanitization: `tenants: [""]` collapses to Any,
    // which matches every non-SubnetLocal candidate including
    // tenant-tagged ones.
    expect(
      a.findNodesScoped({ requireTags: ['gpu'] }, { kind: 'tenants', tenants: [''] }),
    ).toContain(a.nodeId());

    // `tenants: []` (empty list) also falls back to Any.
    expect(
      a.findNodesScoped({ requireTags: ['gpu'] }, { kind: 'tenants', tenants: [] }),
    ).toContain(a.nodeId());
  });

  it('findNodesScoped: tenants list with mixed empty and real ids drops empties', async () => {
    // P2 partial-cleaning case: `tenants: ["", "oem-123"]`
    // sanitizes to `Tenants(["oem-123"])` — real tenant filter
    // semantics preserved, empty entry silently dropped.
    const a = await MeshNode.create({ bindAddr: '127.0.0.1:0', psk: PSK });
    nodes.push(a);

    await a.announceCapabilities({
      tags: ['gpu', 'scope:tenant:oem-123'],
    });

    expect(
      a.findNodesScoped(
        { requireTags: ['gpu'] },
        { kind: 'tenants', tenants: ['', 'oem-123'] },
      ),
    ).toContain(a.nodeId());

    // Real tenant filter: empty + non-matching tenant excludes us.
    expect(
      a.findNodesScoped(
        { requireTags: ['gpu'] },
        { kind: 'tenants', tenants: ['', 'corp-acme'] },
      ),
    ).not.toContain(a.nodeId());
  });

  it('findNodesScoped: regions list with only empty strings falls back to Any', async () => {
    // Same shape as the tenants regression but for regions.
    const a = await MeshNode.create({ bindAddr: '127.0.0.1:0', psk: PSK });
    nodes.push(a);

    await a.announceCapabilities({
      tags: ['relay-capable', 'scope:region:eu-west'],
    });

    expect(
      a.findNodesScoped(
        { requireTags: ['relay-capable'] },
        { kind: 'regions', regions: [''] },
      ),
    ).toContain(a.nodeId());

    expect(
      a.findNodesScoped(
        { requireTags: ['relay-capable'] },
        { kind: 'regions', regions: [] },
      ),
    ).toContain(a.nodeId());
  });
});
